### PR TITLE
Handle `Range` requests in `HTTP::StaticFileHandler`

### LINKED ISSUE #8954: HTTP::StaticFileHandler should support 206 Partial Content feature
# Summary

* This feature is very useful, many sites are using this feature.
* I have not used this feature for the time being,
  * so I have not implemented it for the time being.
* For more explanation about this feature, please refer to [References](#References).
* This may require using some `seek` (File).
  * This feature often appears in (E.g. Restful APIs).

## Specifications

* [RFC 7233, section 4.1: `206 Partial Content`](https://tools.ietf.org/html/rfc7233#section-4.1)
* [RFC 7233, section 3.1: `Range`](https://tools.ietf.org/html/rfc7233#section-3.1)
* [RFC 7233, section 4.2: `Content-Range`](https://tools.ietf.org/html/rfc7233#section-4.2)

## Sample

* Request (I.e. Single, `Range`)

```text
GET /audio/... HTTP/1.1
Host: audio-akp-quic-control-spotify-com.akamaized.net
Connection: keep-alive
Pragma: no-cache
Cache-Control: no-cache
Sec-Fetch-Dest: empty
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Safari/537.36
DNT: 1
Accept: */*
Origin: https://open.spotify.com
Sec-Fetch-Site: cross-site
Sec-Fetch-Mode: cors
Referer: https://open.spotify.com/
Accept-Encoding: identity
Accept-Language: zh-TW,zh;q=0.9,en-US;q=0.8,en;q=0.7
Range: bytes=495259-660589
```

* Response (I.e. Single, `Content-Range`)

```text
HTTP/1.1 206 Partial Content
Last-Modified: Thu, 21 Nov 2019 19:49:02 GMT
ETag: "..."
Content-Type: application/octet-stream
Accept-Ranges: bytes
Cache-Control: no-transform, max-age=30745172
Cache-Control: max-age=315360000, no-transform
Date: Fri, 27 Mar 2020 18:53:39 GMT
Connection: keep-alive
Access-Control-Max-Age: 86400
Access-Control-Allow-Headers: range, pragma, cache-control
Access-Control-Allow-Methods: GET
Access-Control-Allow-Origin: *
Expires: Mon, 25 Mar 2030 18:53:39 GMT
Content-Range: bytes 495259-660589/3592935
Content-Length: 165331
```

* Request (I.e. Several, `Range`)

```text
GET http://localhost:8080/traditional/small.dat HTTP/1.1
Host: localhost:8080
Connection: close
Accept: */*
Accept-Encoding: gzip
User-Agent: Jetty/9.4.19.v20190610
Range: bytes=2275-4550,6825-9100,11375-13650,15925-18200
```

* Response (I.e. Several, `Content-Type: multipart/byteranges`, `Content-Range`)

```text
HTTP/1.1 206 Partial Content
Date: Wed, 15 Nov 2015 06:25:24 GMT
Last-Modified: Wed, 15 Nov 2015 04:58:08 GMT
Content-Length: 1741
Content-Type: multipart/byteranges; boundary=String_separator

--String_separator
Content-Type: application/pdf
Content-Range: bytes 234-639/8000

...the first range...
--String_separator
Content-Type: application/pdf
Content-Range: bytes 4590-7999/8000

...the second range
--String_separator--
```

## References

* [206 Partial Content - HTTP | MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/206)
* [Github | eclipse/jetty.project | multipart/byterange output is invalid to RFC7233](https://github.com/eclipse/jetty.project/issues/3916)
* [Laravel | How to support HTTP Byte Serving in file streams? (Posted 5 years ago)](https://laravel.io/forum/09-23-2014-how-to-support-http-byte-serving-in-file-streams)
* [StackOverflow | How to serve http partial content with Golang?](https://stackoverflow.com/questions/36540610/how-to-serve-http-partial-content-with-go)
* [Range - HTTP | MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range)
* [Content-Range - HTTP | MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Range)

### PR DESCRIPTION
**Pull request by jgaskins**
When fetching a large file, such as a media file, that gets picked up by `HTTP::StaticFileHandler`, the handler copies the entire file to the response, even if the client is just trying to fetch metadata about the file. This can make that metadata request take a lot longer and lead to extra bandwidth being used, which can be expensive.

At the moment, this PR does not handle the [`If-Range`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Range) header, which it looks like is used to ensure that multiple `Range` requests to get small chunks of a file will return chunks of the _same_ file, falling back to returning the entire file in case of an `Etag` or `Last-Modified` mismatch.

Fixes #8954

### CODE REVIEW

File: src/http/server/handlers/static_file_handler.cr
Comment (straight-shoota) [inline src/http/server/handlers/static_file_handler.cr:107]: This parsing mechanism looks like it doesn't do proper error handling. It breaks on invalid range values such as `bytes=1` or `bytes=1-2-3`.
Also, there is some duplication between matching the overall (incomplete) format via regex and then splitting into substrings.

File: src/http/server/handlers/static_file_handler.cr
Comment (straight-shoota) [inline src/http/server/handlers/static_file_handler.cr:111]: Why not use a `Range` here? Seems like a perfect fit.
Size calculation is already included for free.

**Review by straight-shoota — COMMENTED** [review]
I'd like to see some additional specs for an empty file and where start byte is in range, but the end byte is not (I suppose it should clamp to the end of data?).

The `Multipart.parse` method is an independent feature. Could you extract that into a separate PR, please? We should try to separate concerns.

**Comment by jgaskins** [comment]
> I'd like to see some additional specs for an empty file and where start byte is in range, but the end byte is not (I suppose it should clamp to the end of data?).

I was sure I *had* written a test for omitting the end of the range. Looks like it got deleted or rewritten into one of the ones I did end up committing.

> The `Multipart.parse` method is an independent feature. Could you extract that into a separate PR, please? We should try to separate concerns.

It was required to be able to parse the response for multiple ranges. This PR can’t exist without it.

**Comment by straight-shoota** [comment]
> It was required to be able to parse the response for multiple ranges. This PR can’t exist without it.

This PR can depend on the other one.
It's only used for the specs, so you could alternatively turn it a spec helper for this PR in order to merge it independently.

File: src/http/server/handlers/static_file_handler.cr
Comment (jgaskins) [reply src/http/server/handlers/static_file_handler.cr:107]: > This parsing mechanism looks like it doesn't do proper error handling. It breaks on invalid range values such as `bytes=1` or `bytes=1-2-3`.

This is not actionable feedback. Define “breaks”. What do you think the expected and actual behaviors are here?

> Also, there is some duplication between matching the overall (incomplete) format via regex and then splitting into substrings.

What do you mean here?

Comment (jgaskins) [reply src/http/server/handlers/static_file_handler.cr:107]: Additionally, how likely is this header to be malformed? How critical is it to get this precisely correct before the feature is merged at all? If it has to be perfect before it gets merged I’ll just close the PR and publish it as a shard.

**Comment by jgaskins** [comment]
> This PR can depend on the other one.

If you would like to go to all that effort yourself of creating multiple PRs and wiring up their dependencies, you’re more than welcome to, but I will not be splitting PRs for the sake of accounting that ultimately makes no difference.

I’ve already spent over a day on this PR already between trying to understand how these requests work, implementing the feature itself, and doing additional testing with a browser, not counting the time I spent on a previous attempt at this several months ago.

**Comment by straight-shoota** [comment]
Sure, no problem. I can take care of these administrative aspects. Thanks for all the work you have already put into it.

I realize that it seems a bit cumbersome to go the extra steps, but we've realized that it's worth it in the long run for the project to have small and self-contained changes. And we'll take the burdon of this extra work (it's not that much actually, though).

**Comment by straight-shoota** [comment]
I extracted the implementation of `MIME::Multipart.parse` into https://github.com/crystal-lang/crystal/pull/12890. This PR should be merged first, then this one is good to go on top.

File: src/http/server/handlers/static_file_handler.cr
Comment (straight-shoota) [reply src/http/server/handlers/static_file_handler.cr:107]: `bytes=1` raises `IndexError` and `bytes=1-2-3` is interpreted the same as `bytes=1-`. I think both kind of errors should cause an HTTP error response.

> > Also, there is some duplication between matching the overall (incomplete) format via regex and then splitting into substrings.
> What do you mean here?

It's just that validation is essentially split between the regex and the string methods. So some of that is redundant. For example the regex would be unnecessary, if the parsing was bit more precise.

Comment (straight-shoota) [reply src/http/server/handlers/static_file_handler.cr:107]: > Additionally, how likely is this header to be malformed? How critical is it to get this precisely correct before the feature is merged at all? If it has to be perfect before it gets merged I’ll just close the PR and publish it as a shard.

I doesn't has to be perfect. But the basics should be covered and I think proper error handling is necessary for that.
As said below, I can take care of finishing it up and I expect this shouldn't take long to be merged.
The base implementation looks solid, just needs some adjustments to handle edge cases / errors properly.

**Comment by asterite** [comment]
I said it before and I'll say it again: the Ruby core team happily merges incomplete or almost good PRs. Then they introduce extra commits (here we'll use extra PRs) with those additional changes. The end result is that contributory are happier, they don't get frustrated, they don't get to do extra work, and as a bonus, they don't get the final word on how things should be. The last one might be a downside to them, but I'm sure the happiness is of contributing to is greater (I contributed to Ruby a couple of times and that was my experience)

**Comment by asterite** [comment]
Like, instead of saying "please do this and that" it's so much easier to do it ourselves with exactly the code that we have in mind.

**Comment by asterite** [comment]
That said, everything is depends on the chosen process. If contribution is a friction, providing a diff of the intended change could also work. That way the rest of the work will surely depend on the core team.

Happy new year everyone!

**Comment by straight-shoota** [comment]
I amended a couple of changes. Major improvement is in the last commits which add specs for edge cases and fixes the implementation to handle them.
@jgaskins Would you care to take a look?

File: src/http/server/handlers/static_file_handler.cr
Comment (Sija) [inline src/http/server/handlers/static_file_handler.cr:107]: This is unreadable, please avoid such convoluted constructs.

```suggestion
          unless range_header = range_header.lchop?("bytes=")
            return context.response.respond_with_status :bad_request
          end
          unless ranges = parse_ranges(range_header, file_info.size)
            return context.response.respond_with_status :bad_request
          end
```

File: src/http/server/handlers/static_file_handler.cr
Comment (Sija) [inline src/http/server/handlers/static_file_handler.cr:127]:
```suggestion
            MIME::Multipart.build(context.response) do |builder|
```

File: src/http/server/handlers/static_file_handler.cr
Comment (Sija) [inline src/http/server/handlers/static_file_handler.cr:160]:
```suggestion
      return unless start = start.to_i64?
```

File: src/http/server/handlers/static_file_handler.cr
Comment (Sija) [inline src/http/server/handlers/static_file_handler.cr:164]:
```suggestion
        return unless finish = finish.to_i64?
```

**Comment by straight-shoota** [comment]
Sorry @Sija I won't delve into an argument about code style.

File: src/http/server/handlers/static_file_handler.cr
Comment (straight-shoota) [reply src/http/server/handlers/static_file_handler.cr:106]: Yeah, I was debating about this as well. It probably doesn't *really* matter much either way, but this suggestion looks good.

**Comment by straight-shoota** [comment]
Going through the RFC again, there are a couple more edge cases missing. I'll adress that later.

**Comment by straight-shoota** [comment]
Yes, it's because I applied one of the suggestions which only changed the spec, not the implementation.

File: src/http/server/handlers/static_file_handler.cr
Comment (HertzDevil) [inline src/http/server/handlers/static_file_handler.cr:169]: From [here](https://www.rfc-editor.org/rfc/rfc9110.html#section-14.1.2-7):

> A client can refer to the last N bytes (N > 0) of the selected representation using a [suffix-range](https://www.rfc-editor.org/rfc/rfc9110.html#rule.suffix-range). If the selected representation is shorter than the specified [suffix-length](https://www.rfc-editor.org/rfc/rfc9110.html#rule.suffix-range), the entire representation is used.

So a better implementation is:

```suggestion
        return unless finish > 0
        start = {file_size - finish, 0_i64}.max
        finish = file_size - 1
```

Comment (straight-shoota) [reply src/http/server/handlers/static_file_handler.cr:169]: Good catch!

I don't think the `return` is necessary, though. If `finish <= 0`, the range would end up decreasing and thus invalid.

File: src/http/server/handlers/static_file_handler.cr
Comment (Sija) [inline src/http/server/handlers/static_file_handler.cr:169]: Assuming that the `file_size` is `0` this will become `-1` IIRC?

File: src/http/server/handlers/static_file_handler.cr
Comment (HertzDevil) [inline src/http/server/handlers/static_file_handler.cr:112]: [The success condition is "less than"](https://www.rfc-editor.org/rfc/rfc9110.html#section-14.1.2-11.1):

```suggestion
          # See https://www.rfc-editor.org/rfc/rfc9110.html#section-14.1.2-11.1
          if ranges.any? { |range| range.begin >= file_info.size }
```

File: src/http/server/handlers/static_file_handler.cr
Comment (HertzDevil) [reply src/http/server/handlers/static_file_handler.cr:169]: Yes, and upon further inspection this is actually a problem because it seems RFC 9110 does not define a way to indicate 0-byte ranges in the [`Content-Range`](https://www.rfc-editor.org/rfc/rfc9110.html#section-14.4-10) field of a 206 Partial Content response:

> A Content-Range field value is invalid if it contains a [range-resp](https://www.rfc-editor.org/rfc/rfc9110.html#field.content-range) that has a [last-pos](https://www.rfc-editor.org/rfc/rfc9110.html#rule.int-range) value less than its [first-pos](https://www.rfc-editor.org/rfc/rfc9110.html#rule.int-range) value, or a [complete-length](https://www.rfc-editor.org/rfc/rfc9110.html#field.content-range) value less than or equal to its [last-pos](https://www.rfc-editor.org/rfc/rfc9110.html#rule.int-range) value.

In particular `Content-Range: bytes 0--1/0` is invalid. However there is also [this](https://www.rfc-editor.org/rfc/rfc9110.html#section-14.2-7):

> A server that supports range requests MAY ignore a [Range](https://www.rfc-editor.org/rfc/rfc9110.html#field.range) header field when the selected representation has no content (i.e., the selected representation's data is of zero length).

So we could just disregard `Range` and reply with 200 for empty files??? If we handle that early we can guarantee that `file_size` here is positive.

File: spec/std/http/server/handlers/static_file_handler_spec.cr
Comment (HertzDevil) [inline spec/std/http/server/handlers/static_file_handler_spec.cr:186]: These specs should also test that the `Content-Range` header has the correct value for single-part requests, or is [absent for multi-part ones](https://www.rfc-editor.org/rfc/rfc9110.html#section-15.3.7.2-1)

File: spec/std/http/server/handlers/static_file_handler_spec.cr
Comment (HertzDevil) [inline spec/std/http/server/handlers/static_file_handler_spec.cr:257]: This is mostly stylistic, but this is considered a spec failure and should use `fail` instead of `raise`, ditto for below

File: src/http/server/handlers/static_file_handler.cr
Comment (HertzDevil) [inline src/http/server/handlers/static_file_handler.cr:183]: `parse_ranges("-", 10_i64)` returns `[0..9]`, but `bytes=-` is not a valid range.

File: src/http/server/handlers/static_file_handler.cr
Comment (HertzDevil) [inline src/http/server/handlers/static_file_handler.cr:193]: This doesn't reject ranges like `bytes=-0` or even `bytes=--2`; should be `if start || finish <= 0`

File: spec/std/http/server/handlers/static_file_handler_spec.cr
Comment (HertzDevil) [inline spec/std/http/server/handlers/static_file_handler_spec.cr:204]: This is no longer a valid Range header, so I'd suggest to move this spec outside this `describe`

File: spec/std/http/server/handlers/static_file_handler_spec.cr
Comment (HertzDevil) [inline spec/std/http/server/handlers/static_file_handler_spec.cr:347]: Then this should also be grouped into the above `describe`

**Review by HertzDevil — APPROVED** [review]

**Pull request merged**
